101-lab/content/16_templates.md

```diff
@@ -0,0 +1,100 @@
+
+1. On the first line, replace  `apiVersion: v1` with `apiVersion: template.openshift.io/v1`
+2. On the second line, add a new field: `kind: Template` and then rename the `items:` field to `objects:`
+3. Remove `kind: List`
+
+Next, add a new field called 'parameters' near the beginning of our yaml file, above  `objects`. Replace `[username]` with your username. 
+
+```
+parameters: 
+  - name: OWNER
+    required: true
+    value: '[username]'
+  - name: APP_NAMESPACE
+    required: true
+    value: 'dev'
+```
+Next, let's search through the file for the places where our username or namespace suffix `-dev` is being used and replace them with `${OWNER}` and `${APP_NAMESPACE}`. This way, in the future we can just make changes to a parameter values and they'll be applied throughout our template. Be sure to check carefully, don't just find and replace all! Be sure to save your `template.yaml` file. 
+
+## Test the template
+
+The `oc process` command is used to process a template into a resource list. The `-f` flag used to indicate that we need to process a file.  
+
+The `oc create` command can be used to create new objects in OpenShift, again using the ``f` flag. 
+
+Below, we can use the pipe character `|` to take the output from the `oc process` command and feed it into the `oc create` command. We add `--dry run=client` to get a simulated output before actually running the command. This way, we can check that our template is working correctly: 
+
+`oc -n [-dev] process -f template.yaml | oc create -f - --dry-run=client`
+
+Your output should look similar to the below: 
+```
+oc -n d8f105-dev process -f template.yaml | oc apply -f - --dry-run=client  
+
+deployment.apps/rocketchat-mattspencer configured (dry run)
+deploymentconfig.apps.openshift.io/mongodb-mattspencer configured (dry run)
+route.route.openshift.io/rocketchat-mattspencer configured (dry run)
+service/mongodb-mattspencer configured (dry run)
+service/rocketchat-mattspencer configured (dry run)
+configmap/rocketchat-mattspencer-configmap configured (dry run)
+persistentvolumeclaim/mongodb-mattspencer-file configured (dry run)
+secret/mongodb-mattspencer configured (dry run)
+secret/rocketchat-mattspencer-secret configured (dry run)
+```
+
+If the dry run looks to be producing a successful output, let's move on to test our template. 
+
+## Delete objects, recreate them with the template
+
+Currently, out template would just be creating objects that already exist. Let's create a situation where all of the objects that we'd previously created in the `-dev` namespace have been deleted. 
+
+Before deleting objects, it's also good practice to do a dry run to make sure you're only deleing the objects you intend to: 
+
+`oc -n [-dev] delete deployment,dc,route,service,configmap,pvc,secrets -l app=rocketchat-mattspencer --dry-run=client`
```

Comment:
do you always want to use your deploment name iin here? or probally change it to some parameter